src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -0,0 +1,185 @@
+import cats.effect.syntax.concurrent._
+import cats.implicits._
+import fs2.kafka.internal.syntax._
+import org.apache.kafka.clients.producer.{Callback, Producer, ProducerConfig, RecordMetadata}
+
+import scala.concurrent.ExecutionContext
+
+/**
+  * [[TransactionalKafkaProducer]] represents a producer of Kafka messages specialized
+  * for "read-process-write" streams, with the ability to simultaneously produce
+  * `ProducerRecord`s and commit corresponding `CommittableOffset`s using [[produce]].
+  * Records are wrapped in [[TransactionalMessage]] which allow an arbitrary value, that
+  * is a passthrough, to be included in the result.
+  */
+sealed abstract class TransactionalKafkaProducer[F[_], K, V] {
+
+  /**
+    * Produces the `ProducerRecord`s in the specified [[TransactionalMessage]]
+    * in four steps: first a transaction is initialized, then the records are
+    * placed in the buffer of the producer, then the offsets of the records are
+    * sent to the trancation, and lastly the transaction is completed. The
+    * first returned effect completes when the transaction is completed; the
+    * second effect waits for the records to have actually been sent. Note that
+    * it is very slow to wait for individual records to complete sending, but if
+    * you're sure that's what you want, then simply `flatten` the result from this
+    * function.<br>
+    * <br>
+    * If you're only interested in the passthrough value, and not the whole
+    * [[ProducerResult]], you can instead use [[producePassthrough]] which
+    * only keeps the passthrough value in the output.
+    */
+  def produce[G[+ _], P](
+    message: TransactionalMessage[F, G, K, V, P]
+  ): F[F[ProducerResult[G, K, V, P]]]
+
+  /**
+    * Like [[produce]] but only keeps the passthrough value of the
+    * [[ProducerResult]] rather than the whole [[ProducerResult]].
+    */
+  def producePassthrough[G[+ _], P](
+    message: TransactionalMessage[F, G, K, V, P]
+  ): F[F[P]]
+}
+
+private[kafka] object TransactionalKafkaProducer {
+  private[this] def createProducer[F[_], K, V](
+    settings: ProducerSettings[K, V],
+    transactionalId: String,
+    blockingEc: ExecutionContext
+  )(implicit F: Concurrent[F], C: ContextShift[F]): Resource[F, Producer[K, V]] = {
+    Resource
+      .make[F, Producer[K, V]] {
+        settings.producerFactory
+          .create(settings.withProperty(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId))
+      } { producer =>
+        F.delay {
+            producer.close(settings.closeTimeout.asJava)
+          }
+          .start
+          .flatMap(_.join)
+      }
+      .evalMap { producer =>
+        C.evalOn(blockingEc)(F.delay(producer.initTransactions())).as(producer)
+      }
+  }
+
+  def producerResource[F[_], K, V](
+    settings: ProducerSettings[K, V],
+    transactionalId: String,
+    consumerGroupId: String,
+    blockingEc: ExecutionContext
```

Comment:
`transactionalId` and `blockingEc` could go into a new `TransactionalSettings` type. `consumerGroupId` could probably be a field on `TransactionalMessage` instead of a per-consumer setting. Thoughts?

Reply:
I think we can re-use `ProducerSettings`, since we should get an error on `initTransactions` if we forget to set `transactionalId`. So let's add `withTransactionalId` to `ProducerSettings`. In the `0.20.x` branch, we already have a `ExecutionContext` on `ProducerSettings`, so maybe it's worth rebasing against that branch?

Regarding `consumerGroupId`, I'm thinking maybe we should put it on `CommittableOffset` (and `CommittableOffsetBatch`), so the user only has to specify it once for the consumer and nowhere else.

Reply:
I think tracking the group ID in `CommittableOffset` will definitely work. For `CommittableOffsetBatch`, I'm worried there isn't a good default behavior for:
* Creating a batch from two offsets with different group IDs
* Merging two batches with different group IDs

In the transactional case these mismatches should raise errors, but for non-transactional setups they seem harmless.

My thoughts for a middle ground are:
* Track `consumerGroupId` within `CommittableOffset`
* Rework the builder methods for `TransactionalMessage` to validate that all the group IDs match (and return `F[TransactionalMessage]`)

Reply:
Sounds like a good middle ground. 👍